v1.8 backports 2020-07-15

brb opened the pull request:
v1.8 backports 2020-07-15

 * #12443 -- bpf: clear getpeername{4,6} progs (@tklauser)
 * #12415 -- Replace bpf bitfields with a single u8 field and masks (@kkourt)
 * #12427 -- Improve etcd fail-over scenarios (@tgraf)
 * #12456 -- bpf: Fix BPF masq when running with non-hybrid DSR (@brb)
 * #12289 -- Host Policies can break kube if applied incorrectly.  (@jedsalazar)
 * #12440 -- Fix small CRD issue with toGroups (@lbernail)
 * #12457 -- refactor stringSlice type CLI arguments (@JieJhih)
 * #12446 -- cilium: hostport service map fixes (@borkmann)
 * #12451 -- operator: rate limit GC of security identities (@aanm)
 * #12473 -- cilium: simplify hostport's genServiceMappings (@borkmann)
 * #12191 -- Include BPF FS mount status in sysdump output (@soumynathan)
 * #12470 -- pkg/k8s: use copy of objectmeta when fetching from local stores (@aanm)
 * #12462 -- test: Simplify DNS proxy port / TFTP source port conflict test. (@jrajahalme)
 * #12461 -- datapath/linux: protect against concurrent access in NodeValidateImplementation (@tklauser)
 * #12479 -- docs: Improve github actions release docs (@joestringer)
 * #12474 -- policy/api: Support unmanaged entity in policies (@pchaigno)
 * #12492 -- cilium: hostport follow-up (@borkmann)
 * #12494 -- docs: document hostport accessibility in kube-proxy-free guide (@borkmann)
 * #12480 -- Makefile: Fix render-docs with parallel make jobs (@joestringer)
 * #12481 -- Document kvstore leases, update links (@joestringer)
 * #12153 -- doc: Fix clustermesh documentation to set the correct identityMode (@soumynathan)
 * #12497 -- policy: Fix enforcement status of host when PolicyEnforcement=always (@pchaigno)
 * #12495 -- daemon: Fatal on incompatible host firewall options (@pchaigno)
 * #12245 -- Adds TRACE_TO_NETWORK obs label and trace pkts in to-netdev prog. (@Weil0ng)
 * #12486 -- helm: update notes to include Hubble Relay (@Rolinh)
 * #12526 -- clustermesh: Tidy up services connection on failure to reconnect (@joestringer)
 * #12528 -- hubble: add missing code documentation (@Rolinh)
 * #12517 -- Adds documentation for limiting identity-relevant labels used when evaluating Cilium Identities (@seanmwinn)
 * #12529 -- bpf: explicitly set ttl in tunnel key (@borkmann)
 * #12453 -- Add heartbeat to etcd quorum check (@tgraf)
 * #12532 -- docs(troubleshooting):  Remove bugtool related step with --serve flag (@sayboras)
 * #12524 -- endpoint: Write headerfile under full regeneration (@christarazi)
 * #12559 -- bpf: Fix monitor aggregation for 'from-network' (@joestringer)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 12443 12415 12427 12456 12289 12440 12457 12446 12451 12473 12191 12470 12462 12461 12479 12474 12492 12494 12480 12481 12153 12497 12495 12245 12486 12526 12528 12517 12529 12453 12532 12524 12559; do contrib/backporting/set-labels.py $pr done 1.8; done
```

Commit:
doc: Fix clustermesh documentation to set the correct identityMode

[ upstream commit 8c9539205edc25dc06aebe2d38e6f1dd6320b3cf ]

This PR fixes the documentation issues and as well adds in the option
through helm when 'etcd' is enabled.

Fixes: #12125
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

Commit:
Include BPF FS mount status in sysdump output

[ upstream commit fb82325dd5c73b44e524cebd049b06f9b6d7111a ]

This PR adds in a new cilium bpf command to show the status of
the bpffs filesystem mount.

Fixes: #12057
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

Weil0ng committed:
Adds TRACE_TO_NETWORK obs label and trace pkts in to-netdev prog.

[ upstream commit c470e28a82a913f027229614be1bd18944d04250 ]

Signed-off-by: Weilong Cui <cuiwl@google.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

rolinh committed:
hubble/relay: fix function comment formatting

[ upstream commit c57dc9b39273324556af2d43ce921e4d607c2ca0 ]

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

Commit:
Host Policy updates to avoid breaking Kube

[ upstream commit 31592c5ba82a03b782b97a7d2621ee4788589770 ]

Host Policies can break kube if applied incorrectly. If a user unintentionally
applies an overly aggressive host-level netpolicy it will break access to their
nodes and break kube. Given the blast radius of a mistake and the undocumented
fix, I discussed an initial solution in the docs with @pchaigno and we agreed
that a good start is preventing an exmaple policy from loading until the user
changes example values.

This has the benefit of explicitly requiring some thought from users before
applying a policy that could lead to a k8s outage.

Happy to discuss alternative ideas here but the goal is to protect operators
from breaking their environments while testing this feature.

Signed-off-by: Jed Salazar <jed@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

kkourt committed:
bpf: change lb{4,6}_service flags to a u8 field

[ upstream commit a091ba1b4e5bfc96f1c5d9dc77c9c5faa80a0633 ]

This avoids complications of the compiler reordering the fields which we
also use from Go code, and also allows the use of masks in checks.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
etcd: Shuffle list of etcd endpoints

[ upstream commit b95650b30b46b2a095854d3d0043bdc7066aa382 ]

Shuffling the list of etcd endpoints on each bootstrap has two positive
effects:
 * Agents in the cluster will connect to different etcd members.
 * In case etcd fails to fail-over to due to a bug, a Cilium agent
   restart has a chance to fail-over to a healthy etcd member.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
etcd: Fail kvstore status if >= 3 consecutive quorum errors are observed

[ upstream commit 5875b7761b14cdfb1c938f747e927924f7e3c835 ]

In general, the etcd client library should fail over to a healthy etcd
endpoint and quorum errors should automatically resolve. If this does
not happen for some reason, report unhealthy status to trigger a restart
of the agent.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
etcd: Report number of consecutive errors

[ upstream commit 53c17cd91fbabfa4490e367862b6c5a9f238ff1d ]

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
etcd: Return error in isConnectedAndHasQuorum()

[ upstream commit 414cadf475ac965b3efdbafdd318ac432fbb8b34 ]

Instead of just a boolean, return the reason of failure in
isConnectedAndHasQuorum()

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

lbernail committed:
Fix small CRD issue with toGroups

[ upstream commit 678e7f0205ee7f50566d3c2fdd0bd6a3a63bc54c ]

Fixes: f0049da61f4f ("pkg/k8s: fix all structural issues with CNP validation")

Signed-off-by: Laurent Bernaille <laurent.bernaille@datadoghq.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: generalize wildcard lookup for range inclusion/exclusion

[ upstream commit f6ffe0dac6bb518ae9772ef9d5d8146eb776eacf ]

Name it generically and add another bool that allows to control whether
the dst port must be in the nodeport range or outside of it. For the
current users, it is set so that they match on the configured NodePort
range.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: prep datapath for loopback/local address exposure on hostport

[ upstream commit 48bf28aef0f59b09a435c3e66d7dbf641ace234f ]

Prepare the socket LB for a hostport-related wildcard lookup. The NodePort
and HostPort range are exclusive, so the lookup is still fast given we can
bail out early w/o map lookup in sock{4,6}_wildcard_lookup(). Aside from
the range, we must match on HOST_ID dst address for the hostport lookup.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
pkg/kvstore: do not GC keys if they being frequently reused

[ upstream commit 651199ca0ccba59b375c3e484ae4976131d52f15 ]

Some keys can be reused in between GC function calls. To avoid them
being GCed we should not mark them as stale keys and wait for the next
GC call to be executed.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
operator: propagate context for delete identity

[ upstream commit fe9ca34e15273b51c2de5ed22d87557c1dd225ff ]

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: do not answer to surrogate entries from external

[ upstream commit 75556322ff5b5040fc3e61df66c653d43ebe182a ]

Ensure that we don't process such packets from outside if they
ever land on our node. Instead of doing a more expensive memcmp()
on the address, just include this into the test for the underlying
type. The latter is still present for the case where the agent
is restarted & reconfigured and map entries are still stale before
k8s resync.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tklauser committed:
bpf: clear getpeername{4,6} progs

[ upstream commit c8ae6ed00a2a0f518961527b3ac638d1897c944a ]

Attempt to clear previously loaded getpeername progs in bpf/init.sh in
case $HOSTLB is disabled.

Fixes: f9c2d70950cf ("bpf: implement getpeername for sock lb programs")
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
cilium: fix host port handling in kube-proxy free mode

[ upstream commit 7651ecd13ebbfd57a1064dc083b437d16d1ff701 ]

In order to be compatible with portmap's HostPort mode, we need to expose the
host port on all addresses for the case where no hostIP was specified. This also
includes the loopback address. For the case where hostIP was specified, we only
expose the user provided IP address under HostPort.

A typical portmap rule looks like (removing CNI-HOSTPORT-SETMARK from here):

  [...]
  [6:408] -A PREROUTING -m addrtype --dst-type LOCAL -j CNI-HOSTPORT-DNAT
  [...]
  [0:0] -A CNI-DN-1ba85aa0447ede8ba6dab -d 8.8.8.8/32 -p tcp -m tcp --dport 999 -j DNAT --to-destination 10.0.0.217:80
  [0:0] -A CNI-DN-489a67438eed03d272897 -d 127.0.0.1/32 -p tcp -m tcp --dport 298 -j DNAT --to-destination 10.0.0.27:80
  [2:120] -A CNI-DN-526ac7f749f8260cb0d9f -p tcp -m tcp --dport 666 -j DNAT --to-destination 10.0.0.188:80
  [2:120] -A CNI-DN-81de08afe40acdfde3fac -d 192.168.178.29/32 -p tcp -m tcp --dport 777 -j DNAT --to-destination 10.0.0.138:80
  [2:120] -A CNI-HOSTPORT-DNAT -p tcp -m comment --comment "dnat name: \"portmap\" id: \"dd111758b3cc7d46e0c027725f4c1cc3d5ac349c6cb1d6ebb919a5f433f155aa\"" -m multiport --dports 666 -j CNI-DN-526ac7f749f8260cb0d9f
  [...]

In above example, hostPort 666 has no hostIP specified; hostPort 777 had the hostIP
of 192.168.178.29 specified. In such case access via 127.0.0.1:777 is not possible.
hostPort 298 had 127.0.0.1 as hostIP specified, so it's only available via loopback
but not from outside world. And hostPort 999 had 8.8.8.8 as hostIP specified. It's
not forbidden but it won't match/xlate if 8.8.8.8 is not a local address on the node.
Implement these semantics also for the BPF-based replacement by reworking the k8s
watcher's genServiceMappings().

Fixes: #12116
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
pkg/rate: add rate limiter package to rate limit requests

[ upstream commit 08920869151005e1d300eab703b00941eb800055 ]

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
operator: rate limit GC of security identities

[ upstream commit ea57e36b58f0929b4d69300e2ad593b82062d9c8 ]

To prevent bursts of security identities from being deleted in the
KVStore, possibly causing Cilium agent to have a high CPU usage due
policy calculation, this commit adds a rate limiter for such KVStore
deletes. For example, in case there are 1000 identities to GCed, the
operator will delete 250 every minute until all 1000 identities are
GCed.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Fix BPF masq when running with non-hybrid DSR

[ upstream commit d7e2076fc4d632e41154450843dba6559b1926ab ]

Previously, BPF masquerading could not be performed when running with
non-hybrid DSR due to a call to nodeport_nat_fwd() being compiled out.
The latter is responsible for SNATing packets from pods to world.

Fix: 0962c02984 ("datapath: Enable BPF MASQ for veth mode in IPv4")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
cilium: use IPv4zero for v4 surrogate

[ upstream commit 81353c684f2d3a2719def4d79d83fdca63732e89 ]

In IPv6 case we already use net.IPv6zero in LoadBalancerNodeAddresses().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
cilium: push routable flag into service map from agent

[ upstream commit 1550a369736a4ccda5d641acaf10e14bd44ceb96 ]

Final one to enforce drop of surrogate entries for external traffic.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
test: extend hostPort test cases also to loopback and v4-in-v6

[ upstream commit 56e0e7235a70800b65ae0b451886e9b86c14aa84 ]

Extend the batch of hostPort tests to improve coverage.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tklauser committed:
datapath/linux: protect against concurrent access in NodeValidateImplementation

[ upstream commit d88d5f95f3ff0c19fd926abf01ad3f4ee449bf3a ]

The linuxNodeHandler.neighByNode map is also protected by
linuxNodeHandler.mutex in all other code paths. Protect access to it in
(*linuxNodeHandler).NodeValidateImplementation as well.

Fixes #12460

Fixes: 6c06c51926bc ("node: Remove permanent ARP entry when remote node is deleted")
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

jrajahalme committed:
test: Simplify DNS proxy port / TFTP source port conflict test.

[ upstream commit 657295ccbdc5636563206f5e16a67f693b3a8fe5 ]

Since we address NodePort in k8s2 using the DNS proxy port of k8s2 as
the source port from k8s1, one round is enough regardless of the
backend selection, as in both cases the replies are reverse NATted at
k8s2 (where the port conflict was happening before it was fixed by
#12248).

Fixes: #12336
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

LexCC committed:
refactor stringSlice type CLI arguments

[ upstream commit 9fe5b33a55decb598b78a5d3a8706c53c03782a0 ]

Signed-off-by: JieJhih Jhang <jiejhihjhang@gmail.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm committed:
pkg/k8s: use copy of objectmeta when fetching from local stores

[ upstream commit 3a311ab0b0c5698e130e82ce4dcf75efb9d18fe8 ]

This commit fixes a bug introduced by 9975bba1637c where we were
accidentally writing into the object metadata of the local pod store,
which should never happen. Doing that could cause k8s update pod
events to give to the Update function handlers an old pod structure with
such fields modified.

Fixes: 9975bba1637c ("pkg/endpoint: fetch pod and namespace labels from local stores")

Reported-by: Deepesh Pathak <deepshpathak@gmail.com>
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
cilium: simplify hostport's genServiceMappings

[ upstream commit 93e48456b8621bd1eb481e37c14b0701760f3773 ]

Simplify the code which handles HostPort's frontend/backend generation
to make it less complex.

Suggested-by: Andre Martins <andre@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno committed:
policy/api: Support unmanaged entity in policies

[ upstream commit 5e9c427e4d3950d3da60ab066ff47397af472e9c ]

0a9685e introduced the reserved:unmanaged identity, but did not
introduce the associated entity. This commit fixes it and allows the
following in policies:

    - fromEntity:
      - unmanaged

The goal is to allow policies to work even before all pods are restarted
and managed by Cilium. In particular, this can be an issue with host
policies since the host is more likely than pods to need to talk with
unmanaged pods (e.g., on GKE right after deploying Cilium).

Fixes: 0a9685e ("identity: Introduce reserved:unmanaged identity")
Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer committed:
docs: Improve github actions release docs

[ upstream commit 84dda5fe9acc7e184cde891ca15975e82fcbc4f9 ]

We missed this step during the v1.8 release cycle. Clarify the required
steps to hopefully reduce the likelihood we skip past it.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer committed:
Makefile: Fix render-docs with parallel make jobs

[ upstream commit 8f0ee8dd2df66a75be195db70bbd87ea10a77f6d ]

When running make with parallel jobs (make -j N), the ordering between
the build of the html and running the server was not guaranteed. Split
them into separate make commands in the target to ensure the build
happens first.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer committed:
docs: Update kvstore godocs links

[ upstream commit 9369df9577a578011c8053df1b352467a761ee52 ]

godoc.org now points visitors towards pkg.go.dev, so we may as well
point to the new site directly rather than pointing to the old one.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer committed:
docs: Document kvstore lease TTLs

[ upstream commit c4b2c1e6c7245bdaf2bb4a904ed80a4087623f4d ]

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

rolinh committed:
helm: update notes to include Hubble Relay

[ upstream commit affdaa821b0c3f79b7fd36f07bbd277aaa6acb8c ]

This commit updates the NOTES.txt template to ensure that the user is
informed when Hubble Relay is installed be it with or without Hubble UI.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
cilium: ignore explicit hostPort's hostIP when being loopback

[ upstream commit 7e6a7bbb7a086cc2d91328d4534a84ea1596dc31 ]

Do not convert the loopback address into a surrogate, but instead, if
explicitly selected, ignore it and dump a warning message to the agent
log. Converting to surrogate is not comparable given this would also
allow _local_ Pods to connect via the Node's IP.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno committed:
daemon: Fatal if host fw is enabled w/o remote node identities

[ upstream commit 9654cb4562d78f4a03d14ac879c500631e2f7f6a ]

The host firewall currently only works if remote node identities are
distinct from the host identity (i.e., REMOTE_NODE_ID is supported).

It's not clear if this restriction can be removed in the future,
especially once #12345 is merged. In particular, #12345 will redirect
packets from local endpoints through the host device if the destination
security ID is HOST_ID. If the HOST_ID includes remote nodes, then this
redirection is likely to result in packet drops.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
docs: document hostport accessibility in kube-proxy-free guide

[ upstream commit 2c0da38158bd75b48f9606d70fc9b8a47014a5bf ]

Document the expected behavior with regards to addressing.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno committed:
daemon: Fatal if host fw is enabled with per-endpoint routes

[ upstream commit 8d2311424b63b4f49eff8729cd29b724736489d5 ]

The host firewall lives only in the host device and the native devices.
Therefore, we must ensure that packet to and from local pods are routed
through the host device. If not, the host firewall will be bypassed.

Making the host firewall compatible with per-endpoint routes would
require substantial changes, which are unlikely to be a good fit for
backport. In particular, the host policies would need to be accessed
from the bpf_lxc programs. Such changes would impact the size and
complexity of bpf_lxc, as well as the loading of maps.

Per-endpoint routes is a recommended setting in some cases (e.g., GKE).
In such cases, it would be possible to have a partial host firewall
enforcement, with all traffic to and from pods whitelisted. The host
firewall would then only enforce policies on packets to and from world on
the native devices (likely the first use case for the host firewall).

This commit does not implement this second approach and prefers to forbid
per-endpoint routes with the host firewall because a partial enforcement
would likely be confusing to users.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno committed:
policy: Fix enforcement status of host when PolicyEnforcement=always

[ upstream commit 6d691b63436d09544ec3ffef19b4ea7a9d8ad1e1 ]

3a9a353 updated computePolicyEnforcementAndRules() to display the host
endpoint has having no policies enforced as long as the host firewall is
disabled. It however did so only when PolicyEnforcement=default. This
commit fixes it to mark the host policies are not enforced regardless of
PolicyEnforcement's value.

Fixes: 3a9a353 ("policy: Fix enforcement status for host endpoint")
Fixes: f9c205d ("pkg/policy: Host network policies")
Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

christarazi committed:
endpoint: Write headerfile under full regeneration

[ upstream commit 21783d5996019f6a3fb175592129f15339630177 ]

Previously, regenerating an endpoint manually via CLI resulted in
compilation errors as seen below. These errors occurred because the
endpoint headerfile (ep_config.h or lxc_config.h) was not written to the
`<endpoint ID>_next` directory (will be reffered to as "next"), where
the BPF compilation takes place.

The reason the headerfile was not written to the "next" directory was
because Cilium only wrote the headerfile if it was changed (via hash).
However, a manual regeneration triggered through the API, sets the
regeneration level to "compile+load" (a full regeneration), where Cilium
expects all relevant files, including the headerfile, to be present in
the "next" directory (`<endpoint ID>_next`). Hence, the compilation
errors.

This commit fixes this issue by checking whether there has been a
request for a full regeneration, in which case, we write the headerfile
to the "next" directory.

```
root@k8s2:/var/run/cilium/state# clang -emit-llvm -O2 -target bpf -std=gnu89 -nostdinc -D__NR_CPUS__=2 -Wall -Wextra -Werror -Wshadow -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-gnu-variable-s
ized-type-not-at-end -Wdeclaration-after-statement -I/var/run/cilium/state/globals -I1285_next -I/var/lib/cilium/bpf -I/var/lib/cilium/bpf/include -c /var/lib/cilium/bpf/bpf_lxc.c -o /tmp/c
In file included from /var/lib/cilium/bpf/bpf_lxc.c:22:
/var/lib/cilium/bpf/lib/icmp6.h:50:29: error: use of undeclared identifier 'NODE_MAC'
        union macaddr smac, dmac = NODE_MAC;
                                   ^
/var/lib/cilium/bpf/lib/icmp6.h:359:30: error: use of undeclared identifier 'NODE_MAC'
                union macaddr router_mac = NODE_MAC;
                                           ^
In file included from /var/lib/cilium/bpf/bpf_lxc.c:26:
/var/lib/cilium/bpf/lib/lxc.h:67:29: error: use of undeclared identifier 'NODE_MAC'
        union macaddr router_mac = NODE_MAC;
                                   ^
/var/lib/cilium/bpf/bpf_lxc.c:159:10: error: implicit declaration of function 'lookup_ip6_remote_endpoint' [-Werror,-Wimplicit-function-declaration]
                info = lookup_ip6_remote_endpoint(&orig_dip);
                       ^
/var/lib/cilium/bpf/bpf_lxc.c:159:10: note: did you mean 'lookup_ip6_endpoint'?
/var/lib/cilium/bpf/lib/eps.h:13:1: note: 'lookup_ip6_endpoint' declared here
lookup_ip6_endpoint(struct ipv6hdr *ip6)
^
/var/lib/cilium/bpf/bpf_lxc.c:159:8: error: incompatible integer to pointer conversion assigning to 'struct remote_endpoint_info *' from 'int' [-Werror,-Wint-conversion]
                info = lookup_ip6_remote_endpoint(&orig_dip);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/cilium/bpf/bpf_lxc.c:529:10: error: implicit declaration of function 'lookup_ip4_remote_endpoint' [-Werror,-Wimplicit-function-declaration]
                info = lookup_ip4_remote_endpoint(orig_dip);
                       ^
/var/lib/cilium/bpf/bpf_lxc.c:529:10: note: did you mean 'lookup_ip4_endpoint'?
/var/lib/cilium/bpf/lib/eps.h:35:1: note: 'lookup_ip4_endpoint' declared here
lookup_ip4_endpoint(const struct iphdr *ip4)
^
/var/lib/cilium/bpf/bpf_lxc.c:529:8: error: incompatible integer to pointer conversion assigning to 'struct remote_endpoint_info *' from 'int' [-Werror,-Wint-conversion]
                info = lookup_ip4_remote_endpoint(orig_dip);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/cilium/bpf/bpf_lxc.c:1027:10: error: implicit declaration of function 'lookup_ip6_remote_endpoint' [-Werror,-Wimplicit-function-declaration]
                info = lookup_ip6_remote_endpoint(src);
                       ^
/var/lib/cilium/bpf/bpf_lxc.c:1027:8: error: incompatible integer to pointer conversion assigning to 'struct remote_endpoint_info *' from 'int' [-Werror,-Wint-conversion]
                info = lookup_ip6_remote_endpoint(src);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/cilium/bpf/bpf_lxc.c:1256:10: error: implicit declaration of function 'lookup_ip4_remote_endpoint' [-Werror,-Wimplicit-function-declaration]
                info = lookup_ip4_remote_endpoint(ip4->saddr);
                       ^
/var/lib/cilium/bpf/bpf_lxc.c:1256:8: error: incompatible integer to pointer conversion assigning to 'struct remote_endpoint_info *' from 'int' [-Werror,-Wint-conversion]
                info = lookup_ip4_remote_endpoint(ip4->saddr);
                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11 errors generated.
```

Fixes https://github.com/cilium/cilium/pull/10630
Fixes https://github.com/cilium/cilium/issues/12005

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Restart etcd connection on quorum errors

[ upstream commit e88a9093a58f5fbd89fbcbfc39ebd4fde45e8d14 ]

Watch the status of the etcd conection and restart the connection if
quorum loss is detected. Given that lock acquisition is disabled for
clustermesh, the quorum check equals to the ability to receive updates
on the heartbat key.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Fix comments on scope of rc.mutex

[ upstream commit 8599664980a52b1a8e18dc782aa6f210e87c18a5 ]

Reported-by: @sayboras
Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Disable initlock quorum check

[ upstream commit 645355561bf3c8fa628a2053a7276ca03c73937a ]

Clustermesh is never performing write operations so the lock-based
quorum check is only adding contention to remote etcds.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
doc: Document etcd failure behavior

[ upstream commit 4d831cd7934e732ab0006d114fc7977081b9619a ]

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Improve error message for inital connection attempt

[ upstream commit cc162e5e7322855cb8deadc7f8fa1b5020757a10 ]

"Backend not initialized" does not mean much to users.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
kvstore: Add metric to count quorum errors

[ upstream commit 8f0b57d0bae50208dfc87be0b92e95485480cd83 ]

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
kvstore: Add hearbeat to detect stale etcd connections

[ upstream commit d61fdcd537751eedf6f24c820ed8568d386492a9 ]

Adds a heartbeat written to a key in an interval (1min) by the operator.
Each etcd client installs a watcher to watch the heartbeat key. When the
heartbeat is not updated in 2*interval, the quorum check will start
failing:

```
KVStore:                Ok   etcd: 1/1 connected, lease-ID=29c6732d5d580cb5, lock lease-ID=29c6732d5d580cb7, has-quorum=2m2.778966915s since last heartbeat update has been received, consecutive-errors=1: https://192.168.33.11:2379 - 3.4.9 (Leader)
```

When enough consecutive errors have accumulated, the kvstore subsystem
will start failing:
```
KVStore:                Failure   Err: quorum check failed 8 times in a row: 4m28.446600949s since last heartbeat update has been received
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
kvstore: Improve initial status message

[ upstream commit a7510398179c51d357232ce9eda14d8e8741d0db ]

The initial status message of the etcd subsystem is:

```
KVStore:                Ok   No connection to etcd
```

This can be misleading as it does not indicate whether the etcd session
was ever established or not. Clarify this:

```
KVStore:                Ok   Waiting for initial connection to be established
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

seanmwinn committed:
Adds documentation for limiting identity-relevant labels

[ upstream commit 0945666ac60328d4ffd42b08ccae6e4509d65bde ]

Fixes: #11540
Signed-off-by: Sean Winn <sean@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
kvstore: Log errors while closing etcd client

[ upstream commit 03b3c105bb427d6d062f609fe084021dc766f228 ]

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Release old connection asynchronously

[ upstream commit 0b0dd558fe3df45ec78ddcb13467cad1b8c2d1cc ]

When releasing the etcd connection, sessions are attempted to be
revoked. In the event of an unhealthy etcd connection, the operation
will fail and time out. This operation will take a long time though.
Instead of blocking, release the resources in the background.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tgraf committed:
clustermesh: Report number of failures in status

[ upstream commit 3e951e15ec9dcc92c8b53982713f8af5fe631684 ]

Good condition:
```
   cluster2: ready, 4 nodes, 3 identities, 1 services, 0 failures (last: never)
```

Bad condition:
```
   cluster2: not-ready, 0 nodes, 0 identities, 0 services, 1 failures (last: 9s ago)
```

Signed-off-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

seanmwinn committed:
Changes based on comments in the pull request

[ upstream commit a5fe80eb8bb017d372729a79cf529cfe0e4eefb4 ]

Signed-off-by: Sean Winn <sean@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer committed:
clustermesh: Tidy up services connection on failure

[ upstream commit 7c699cbaae94f7e3148c2fdb7e0f1579c260a5b2 ]

In the case where there is a failure to watch remote identities from the
peer cluster, we would not close out the connection to the services
store properly. Fix it up.

The context usage in this code should also be investigated more closely
but we can fix that up in a subsequent commit.

Fixes: f28f6cd887f6 ("clustermesh: Use remote kvstore backend")
Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann committed:
bpf: explicitly set ttl in tunnel key

[ upstream commit 38d9bf589e6be9db6694dada7c1a4f407098b40f ]

Maciej reported that when using vxlan packets we can see a TTL of 64
in the packet while for geneve packets the TTL is set to 0.

This is a kernel issue in that vxlan driver in its vxlan_xmit_one()
routine derives TTL from the route if otherwise not explicitly set
(such as by BPF tunnel key):

  ttl = ttl ? : ip4_dst_hoplimit(&rt->dst);

In geneve driver however, geneve_xmit_skb() only does the above in
non-collect_md mode, which means, if not explicitly set, the TTL will
remain 0 here.

I'll post a kernel fix separately, but simple workaround is to just
set TTL in BPF tunnel key to fixed value of 64.

tunnel_ttl is part of 4.9 bpf uapi header, so there are no issues with
backwards compat.

Reported-by: Maciej Skrocki <maciejskrocki@google.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

rolinh committed:
hubble/testutils: add missing code documentation

[ upstream commit 7d871b24286b61fc0c976021392850fbdb79dcec ]

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

sayboras committed:
docs(troubleshooting): Remove --serve flag in bugtool

[ upstream commit 23265dd883f87cf6a0973d7c30018e25320427dc ]

The flag --serve is removed in bugtool in PR #6237, hence related docs
should be removed as well.

Signed-off-by: Tam Mach <sayboras@yahoo.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno reviewed (approved):
:+1: for my changes. I also checked changes from @Weil0ng's PR are okay.

borkmann reviewed (approved):
for my PRs

christarazi reviewed (approved):
Looks good for my change

joestringer reviewed (approved):
Backports of my changes LGTM

qmonnet commented on Documentation/install/upgrade.rst:
```diff
@@ -598,6 +600,7 @@ The following metrics have been renamed:
 * ``cilium_operator_eni_resync_total`` to ``cilium_operator_ipam_resync_total``
 * ``cilium_operator_eni_aws_api_duration_seconds`` to ``cilium_operator_ipam_api_duration_seconds``
 * ``cilium_operator_eni_ec2_rate_limit_duration_seconds`` to ``cilium_operator_ipam_api_rate_limit_duration_seconds``
+* ``cilium_operator_ipam_ec2_resync`` to ``cilium_operator_ipam_resync``
```
I think this line sneaked in by mistake in your backport for #12453. The rename comes from  #12413 which has not been backported as far as I can tell.

brb replied:
Good catch! Fixed.

qmonnet reviewed (changes requested):
All backports that have not been reviewed by their authors yet look straightforward with no conflict (or no change at all from what I could see).

I found a single exception below, please fix.

joestringer committed:
bpf: Fix monitor aggregation for 'from-network'

[ upstream commit 037bef5260918fba7fac19971593c80ff020485c ]

Previously, we did not take into account 'from-network' sources in the
monitor aggregation logic check in `send_trace_notify()`, which was fine
because we rarely ever sent such events (limited to ipsec for instance).
However, since commit c470e28a82a9 we also use this in bpf_host which
suddenly means that any and all traffic from the network will trigger
monitor events, flooding the monitor output.

Fixes: 7a4b0beccbfe ("bpf: Add MonitorAggregation option")
Fixes: c470e28a82a9 ("Adds TRACE_TO_NETWORK obs label and trace pkts in to-netdev prog.")
Fixes: https://github.com/cilium/cilium/issues/12555

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

qmonnet reviewed (approved):
Looks all good to me, let's wait for the tests and we're good to go!

tklauser reviewed (approved):
👍 for my changes

aanm reviewed (approved):
LGTM for my changes
